orttraining/orttraining/core/graph/gradient_builder.cc

```diff
@@ -1805,9 +1772,24 @@ IMPLEMENT_GRADIENT_BUILDER(GetPythonOpGradient) {
     input_args.push_back(O(i));
   }
 
+  // src_attrs["input_requires_grads"] stores all inputs's requires_grad attributes,
+  // including both tensor inputs and non-tensor inputs (e.g. constants), here we filter out
+  // those non-tensor inputs when constructing PythonOpGrad's outputs.
```

Comment:
input_requires_grads is generated during export, right?  If so, do we have a mechanism to detect if the requirements change during training -- say if, for some reason, the graph is initially exported with more/fewer tensors requiring gradients than?  [ Not a use case I have seen, but looking to catch this kind of issue and report and error if it does occur ]

Reply:
I understand the question this way (correct me if I am wrong): we need check the tensor + non-tensor inputs counts , for all attributes like "input_requires_grads , input_convention". For this, we have some checks when PythonOp is constructed. And I just added more checks enhance it.